crates/binjs_io/src/context/varnum.rs

```diff
@@ -0,0 +1,213 @@
+//! Implementation of varnums used in Context-0.x.
+//!
+//! Note that this is NOT the same implementation as used in other experiments.
+//!
+//! # Format
+//!
+//! ```
+//! varu32 ::= [0b1nnnnnnn]{0..4} 0b0nnnnnnn
+//! ```
+//!
+//! Bits are interpreted as little-endian.
+//! A first bit of `1` indicates a non-last byte.
+//! A first bit of `0` indicates a last byte.
+
+use std::io::{self, Error, Read, Result, Write};
+
+// --------------- Reading
+
+/// A result of reading data from a byte-oriented stream.
+pub struct ByteValue<T> {
+    /// The value read.
+    value: T,
+
+    /// The number of bytes consumed.
+    byte_len: usize,
+}
+
+/// A reader that may read varu32-encoded u32 values from a stream.
+pub trait ReadVaru32 {
+    /// Read a single varu32.
+    ///
+    /// Note that this operation may return denormalized 0 values, e.g.
+    /// `ByteValue { value: 0, byte_len: 5 }`. Such values may be used
+    /// to represent exceptional cases.
+    fn read_varu32_no_normalization(&mut self) -> Result<ByteValue<u32>>;
+}
+
+impl<T> ReadVaru32 for T
+where
+    T: Read,
+{
+    fn read_varu32_no_normalization(&mut self) -> Result<ByteValue<u32>> {
+        let mut result: u32 = 0;
+        let mut shift: u8 = 0;
+        loop {
+            debug_assert!(shift < 32);
+            let mut bytes = [0];
+            self.read_exact(&mut bytes)?;
+
+            let byte = bytes[0];
+            let new_result = result | (((byte & 0b01111111) as u32) << shift);
+            if new_result < result {
+                return Err(Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Overflow during read_varu32_no_normalization",
+                ));
+            }
+
+            result = new_result;
+            shift += 7;
+
+            if byte & 0b10000000 == 0 {
+                return Ok(ByteValue {
+                    value: result,
+                    byte_len: (shift / 7) as usize,
+                });
+            }
+
+            if shift >= 32 {
+                return Err(Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Overflow during read_varu32_no_normalization (too many digits)",
+                ));
+            }
+        }
+    }
+}
+
+#[test]
+fn test_read_varu32_no_normalization_must_work() {
+    use std::io::Cursor;
+    for (input, expected) in &[
+        // Various versions of 0
+        (vec![0], 0),
+        (vec![0b10000000, 0b00000000], 0),
+        (vec![0b10000000, 0b10000000, 0b00000000], 0),
```

Comment:
Maybe also test a couple of multi-length, denormalized non-zero values?

Reply:
Looks like Travis is complaining about cargo fmt.

Reply:
Which values do you have in mind?